src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -240,65 +241,80 @@ private static bool AreCompatible(DynamicallyAccessedMemberTypes serviceDynamica
+                    {
+                        if (TryCreateExact(_descriptors[i], itemType, callSiteChain, slot) is { } callSite)
+                        {
+                            AddCallSite(callSite, i);
                         }
                     }
-                    else
+                    for (int i = _descriptors.Length - 1; i >= 0; i--)
                     {
-                        int slot = 0;
-                        // We are going in reverse so the last service in descriptor list gets slot 0
-                        for (int i = _descriptors.Length - 1; i >= 0; i--)
+                        if (TryCreateOpenGeneric(_descriptors[i], itemType, callSiteChain, slot, throwOnConstraintViolation: false) is { } callSite)
                         {
-                            ServiceDescriptor descriptor = _descriptors[i];
-                            ServiceCallSite? callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot) ??
-                                           TryCreateOpenGeneric(descriptor, itemType, callSiteChain, slot, false);
-
-                            if (callSite != null)
-                            {
-                                slot++;
-
-                                cacheLocation = GetCommonCacheLocation(cacheLocation, callSite.Cache.Location);
-                                callSites.Add(callSite);
-                            }
+                            AddCallSite(callSite, i);
                         }
-
-                        callSites.Reverse();
                     }
 
+                    callSitesByIndex.Sort((a, b) => a.Key.CompareTo(b.Key));
+                    callSites = callSitesByIndex.Select(p => p.Value).ToArray();
 
-                    ResultCache resultCache = ResultCache.None;
-                    if (cacheLocation == CallSiteResultCacheLocation.Scope || cacheLocation == CallSiteResultCacheLocation.Root)
+                    void AddCallSite(ServiceCallSite callSite, int index)
                     {
-                        resultCache = new ResultCache(cacheLocation, callSiteKey);
+                        slot++;
+
+                        cacheLocation = GetCommonCacheLocation(cacheLocation, callSite.Cache.Location);
+                        callSitesByIndex.Add(new(index, callSite));
                     }
+                }
 
-                    return _callSiteCache[callSiteKey] = new IEnumerableCallSite(resultCache, itemType, callSites.ToArray());
+                ResultCache resultCache = ResultCache.None;
+                if (cacheLocation == CallSiteResultCacheLocation.Scope || cacheLocation == CallSiteResultCacheLocation.Root)
+                {
+                    resultCache = new ResultCache(cacheLocation, callSiteKey);
                 }
 
-                return null;
+                return _callSiteCache[callSiteKey] = new IEnumerableCallSite(resultCache, itemType, callSites);
             }
```

Comment:
Note: with this change, I think we could restore validation of open generic services which was removed here: https://github.com/dotnet/runtime/pull/68053/files#diff-b9824a372eeae2c312b58b219349b8b0e4612d432c122f87b3ef45850ec2c7c2L135

Thoughts?